python_modules/libraries/dagster-dbt/dagster_dbt/cli/resources_v2.py

```diff
@@ -130,6 +131,46 @@ def get_asset_key_by_output_name(self, output_name: str) -> AssetKey:
         """Get a dbt node's default asset key by its Dagster output name."""
         return self.node_info_to_asset_key(self.get_node_info_by_output_name(output_name))
 
+    def get_asset_key_for_dbt_unique_id(self, unique_id: str) -> AssetKey:
+        node_info = self.node_info_by_dbt_unique_id.get(unique_id)
+
+        if not node_info:
+            raise DagsterInvalidInvocationError(
+                f"Could not find a dbt node with unique_id: {unique_id}. A unique ID consists of"
+                " the node type (model, source, seed, etc.), project name, and node name in a"
+                " dot-separated string. For example: model.my_project.my_model"
+            )
+
+        return self.node_info_to_asset_key(node_info)
+
+    def get_asset_keys_for_dbt_source(self, source_name: str) -> Sequence[AssetKey]:
```

Comment:
how can a source end up with multiple asset keys?

Reply:
A dbt source is referenced by a `source_name` and a `table_name`: https://docs.getdbt.com/reference/dbt-jinja-functions/source. A source can have multiple tables.

Reply:
Ah I see. What's the scenario where we'd imagine someone using this method? Also, this reminds me that the new methods should have docstring.

Reply:
A source (e.g. `fivetran`) could have multiple tables that are synced by one API call. So this would require the user to define a `@multi_asset` for the source.

Reply:
I'm still not entirely following. When would someone be defining dependencies between non-dbt Dagster assets and dbt sources? I thought the whole point of dbt sources was to make source data available to dbt models.

That said, it seems like including this method is fairly low cost, so not a big deal if we don't have a super concrete use case for it. Mostly just curious.

Reply:
This method has a concrete use case and it is to define software-defined assets for the dbt sources.

Like you said, the point of dbt sources is to define the source data for the dbt models. In dbt, this defines the upstream dependencies (i.e. we map the `(source_name, table_name)` to an `AssetKey`) of the dbt models that are external to the system.

So, the dependencies are already defined when the dbt project is read by the `dagster-dbt` integration. However, to materialize these dbt sources in Dagster, the user needs to define the logic associated with these upstream `AssetKey`'s.

This function, `get_asset_keys_for_dbt_source`, gives you the `AssetKey`'s that you need to define the software-defined assets for the dbt sources.